llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,435 @@
+                                   SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs) {
+  LLVM_DEBUG(dbgs() << " WAW RegRewriter: Register to replace "
+                    << TRI->getName(VRM->getPhys(Reg)) << "\n");
+  SmallVector<MCPhysReg, 16> Replacements =
+      getPriorityReplacementRegs(Reg, ForbiddenPhysRegs);
+
+  for (const MCPhysReg &PhysReg : Replacements) {
+
+    const LiveInterval &LI = LIS->getInterval(Reg);
+    LiveRegMatrix::InterferenceKind IK = LRM->checkInterference(LI, PhysReg);
+
+    if (IK == LiveRegMatrix::IK_Free) {
+      LLVM_DEBUG(dbgs() << "     WAW Replacement: Virtual Register "
+                        << Register::virtReg2Index(Reg) << " will replace "
+                        << printReg(VRM->getPhys(Reg), TRI, 0, MRI) << " with "
+                        << printReg(PhysReg, TRI, 0, MRI) << '\n');
+      LRM->unassign(LI);
+      LRM->assign(LI, PhysReg);
+      addAliasRegs(ForbiddenPhysRegs, PhysReg);
+      return true;
+    }
+  }
+  return false;
+}
+
+SmallVector<MCPhysReg, 16> AIEWawRegRewriter::getPriorityReplacementRegs(
+    const Register Reg, const SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs) {
+  SmallVector<MCPhysReg, 16> PhysRegs;
+  const TargetRegisterClass *RC = MRI->getRegClass(Reg);
+
+  for (const MCPhysReg &PhysReg : RC->getRegisters()) {
+    // iterate Regs of RC and allocate first found reg that does not
+    // interfere
+    if (find(ForbiddenPhysRegs, PhysReg) != ForbiddenPhysRegs.end())
+      continue;
+
+    LiveRegMatrix::InterferenceKind IK =
+        LRM->checkInterference(LIS->getInterval(Reg), PhysReg);
+    if (IK == LiveRegMatrix::IK_Free)
+      PhysRegs.push_back(PhysReg);
+  }
+
+  LLVM_DEBUG(dbgs() << "Renaming Candidate " << printReg(Reg, TRI, 0, MRI)
+                    << " Ordering";
+             for (const MCPhysReg &PhysReg
+                  : PhysRegs) {
+               dbgs() << " " << printReg(PhysReg, TRI, 0, MRI);
+             } dbgs()
+             << "\n";);
+
+  return PhysRegs;
+}
+
+bool AIEWawRegRewriter::isIdentityCopy(const MachineInstr &MI) {
+  if (!MI.isCopy())
+    return false;
+
+  const Register DstPhyReg = getPhysReg(MI.getOperand(0));
+
+  const Register SrcPhyReg = getPhysReg(MI.getOperand(1));
+
+  return DstPhyReg == SrcPhyReg;
+}
+
+bool AIEWawRegRewriter::isUndefReserved(
+    const MachineOperand &Def, const MachineInstr &MI,
+    const MachineBasicBlock &MBB,
+    std::map<unsigned, SmallVector<const MachineInstr *, 4>> &UndefPartReuse) {
+  unsigned ID = Def.getReg().id();
+  if (Def.isUndef() && UndefPartReuse.find(ID) == UndefPartReuse.end()) {
+    UndefPartReuse[ID] = SmallVector<const MachineInstr *, 4>();
+  }
+
+  if (UndefPartReuse.find(ID) == UndefPartReuse.end())
+    return false;
+
+  UndefPartReuse[ID].push_back(&MI);
+
+  // Undef Reg is still being filled, too early to tell, if it should be
+  // removed
+  if (Def.isUndef() && !isLastDef(Def, MI, MBB))
+    return true;
```

Comment:
Note: `undef` really only appears on the first partial def. If a register is defined through e.g. 4 sub-regs, then only the first assignment will get undef, not the "middle ones". I guess what you want to check here is rather `if (Def.getSubReg() && !isLastDef(...))`.

But as said in https://github.com/Xilinx/llvm-aie/pull/168/files#r1743706575, I wonder if we could not just prevent renaming altogether when a vreg is defined through a subreg.

Reply:
I rewrote the whole for clarity.